src/H5VLint.c

```diff
@@ -1003,6 +1003,33 @@ H5VL_conn_dec_rc(H5VL_t *connector)
     FUNC_LEAVE_NOAPI(ret_value)
 } /* end H5VL_conn_dec_rc() */
 
+/*-------------------------------------------------------------------------
+ * Function:    H5VL__conn_dest
+ *
+ * Purpose:     Destroys a VOL connector struct. This should be used
+ *              directly when cleaning up a VOL connector with no references.
+ *
+ * Return:      SUCCEED/FAIL
+ *
+ *-------------------------------------------------------------------------
+ */
+herr_t
+H5VL_conn_dest(H5VL_t *connector)
+{
+    herr_t ret_value = SUCCEED; /* Return value */
+
+    FUNC_ENTER_NOAPI(FAIL)
+
+    if (H5I_dec_ref(connector->id) < 0)
+        HGOTO_ERROR(H5E_VOL, H5E_CANTDEC, FAIL, "unable to decrement ref count on VOL connector");
+    H5FL_FREE(H5VL_t, connector);
+
+    connector = NULL;
```

Comment:
Setting the connector to NULL on free is new behavior, but it's necessary to avoid potential attempts to double free the connector.

Specifically within `H5VL_register_using_vol_id()` -> `H5VL_register()` -> `H5VL__new_vol_obj()`.  `H5VL__new_vol_obj()` is provided with a connector with zero references and attempts to create a new object. If it fails after incrementing the nrefs of the connector, it uses `H5VL_conn_dec_rc`, which frees the connector upon brings it to `nrefs = 0`. If the connector isn't set to NULL, then `H5VL_register_using_vol_id()` would attempt to double free it.

Reply:
Setting connector to NULL has only a local affect and therefore is unnecessary here.

Reply:
Good catch. This still needs to be resolved somehow, but it looks like it will be a little involved and can be done in a separate PR.